includes/class-wc-stripe-intent-controller.php

```diff
@@ -762,6 +762,85 @@ public function create_and_confirm_payment_intent( $payment_information ) {
 		return $payment_intent;
 	}
 
+	/**
+	 * Updates and confirm a payment intent with the given payment information.
+	 * Used for dPE.
+	 *
+	 * @param object $payment_intent       The payment intent to update.
+	 * @param array $payment_information The payment information needed for creating and confirming the intent.
+	 *
+	 * @throws WC_Stripe_Exception - If the create intent call returns with an error.
+	 *
+	 * @return array
+	 */
+	public function update_and_confirm_payment_intent( $payment_intent, $payment_information ) {
+		// Throws a WC_Stripe_Exception if required information is missing.
+		$required_params = [
+			'capture_method',
+			'payment_method',
+			'shipping',
+			'selected_payment_type',
+			'payment_method_types',
+		];
+
+		$missing_params = [];
+		foreach ( $required_params as $param ) {
+			// Check if they're set. Some can be null.
+			if ( ! array_key_exists( $param, $payment_information ) ) {
+				$missing_params[] = $param;
+			}
+		}
+
+		$shopper_error_message = __( 'Please reach out to us if the problem persists.', 'woocommerce-gateway-stripe' );
+
+		// Bail out if we're missing required information.
+		if ( ! empty( $missing_params ) ) {
+			throw new WC_Stripe_Exception(
+				sprintf(
+					'The information for creating and confirming the intent is missing the following data: %s.',
+					implode( ', ', $missing_params )
+				),
+				$shopper_error_message
+			);
+		}
+
+		$selected_payment_type = $payment_information['selected_payment_type'];
+		$payment_method_types  = $payment_information['payment_method_types'];
+
+		$request = [
+			'capture_method' => $payment_information['capture_method'],
+			'payment_method' => $payment_information['payment_method'],
+			'shipping'       => $payment_information['shipping'],
```

Comment:
We were using level3 data in the original implementation. Should we include it in the new one as well?

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/abstracts/abstract-wc-stripe-payment-gateway.php#L1520-L1525

Reply:
Not sure. Do you have the updated document for level 3 data? [The one in the code comments](https://stripe.com/docs/level3#level-iii-usage-requirements) is not working anymore. Per [this endpoint documentation on Stripe](https://stripe.com/docs/api/payment_intents/confirm), it does not seem to support the param `level_3`.

Reply:
> Per [this endpoint documentation on Stripe](https://stripe.com/docs/api/payment_intents/confirm), it does not seem to support the param level_3.

As far as I know, it's only displayed in the documentation if the Stripe account you're logged into supports this parameter. This is how it looks when I'm logged into the WooPayments one.

<img width="3334" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/9d01b1a5-fff9-4ff0-bb7e-3f482338ad26">

We have [this method](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-stripe-api.php#L190-L274) that handles whether to send this parameter depending on the merchant's account.

We should probably keep it because that's how it's currently being done (reference in my previous comment), but glad to change it if there's a reason to.

Reply:
Oh! Do you know how to enable that on my test account? Anyway, I included the level 3 data and replaced the request method with this one 👍

Reply:
Thanks!

> Do you know how to enable that on my test account?

I don't, but I can look it up.